drv/grapefruit-seq-server/src/main.rs

```diff
@@ -0,0 +1,266 @@
+            sys_api::Pull::None,
+        );
+        sys.gpio_reset(FAULT_PIN_L);
+
+        // To allow for the possibility that we are restarting, rather than
+        // starting, we take care during early sequencing to _not turn anything
+        // off,_ only on. This means if it was _already_ on, the outputs should
+        // not glitch.
+
+        // To program the FPGA, we're using "slave serial" mode.
+        //
+        // See "7 Series FPGAs Configuration", UG470 (v1.17) for details,
+        // as well as "Using a Microprocessor to Configure Xilinx 7 Series FPGAs
+        // via Slave Serial or Slave SelectMAP Mode Application Note" (XAPP583)
+
+        // Configure the PROGRAM_B line to the FPGA
+        sys.gpio_set(FPGA_PROGRAM_L);
+        sys.gpio_configure_output(
+            FPGA_PROGRAM_L,
+            sys_api::OutputType::OpenDrain,
+            sys_api::Speed::Low,
+            sys_api::Pull::None,
+        );
+
+        // Pulse PROGRAM_B low for 1 ms to reset the bitstream
+        // (T_PROGRAM is 250 ns, so this is fine)
+        // https://docs.amd.com/r/en-US/ds189-spartan-7-data-sheet/XADC-Specifications
+        sys.gpio_reset(FPGA_PROGRAM_L);
+        hl::sleep_for(1);
+        sys.gpio_set(FPGA_PROGRAM_L);
+
+        // Wait for INIT_B to rise
+        loop {
+            let init = sys.gpio_read(FPGA_INIT_L) != 0;
+            ringbuf_entry!(Trace::Ice40Init(init));
+            if init {
+                break;
+            }
+
+            // Do _not_ burn CPU constantly polling, it's rude. We could also
+            // set up pin-change interrupts but we only do this once per power
+            // on, so it seems like a lot of work.
+            hl::sleep_for(2);
+        }
+
+        // Do we have to send the synchronization word ourself, or is it built
+        // into the bitstream?
+        // Same with device ID check
+        // Load bitstream
+        //
+        // SP_TO_FPGA_CFG_CLK / SP_TO_FPGA_CFG_DAT
+        // This is on SPI2, port B
+        //
+        // Wait for DONE (FPGA_TO_SP_CONFIG_DONE)
+
+        // Bind to the sequencer device on our SPI port
+        let seq = spi.device(drv_spi_api::devices::FPGA);
+
+        // TODO do we need to send the bus width / synchronization word / device
+        // ID ourselves, or are they built into the image?
+
+        let blob = aux
+            .get_blob_by_tag(*b"FPGA")
+            .map_err(|_| SeqError::AuxMissingBlob)?;
+        let mut scratch_buf = [0u8; 128];
+        let mut pos = blob.start;
+        let mut sha = Sha3_256::new();
+        let mut decompressor = gnarle::Decompressor::default();
+        while pos < blob.end {
+            let amount = (blob.end - pos).min(scratch_buf.len() as u32);
+            let chunk = &mut scratch_buf[0..(amount as usize)];
+            if let Err(e) = aux.read_slot_with_offset(blob.slot, pos, chunk) {
+                return Err(SeqError::AuxReadError(e));
+            }
```

Comment:
nitpicky, not important: isn't this just:
```suggestion
            aux.read_slot_with_offset(blob.slot, pos, chunk).map_err(SeqError::AuxReadError)?;
```